File: implementation/scipamato/core/persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluator.java
package ch.difty.scipamato.core.persistence.paper.searchorder;

import org.jetbrains.annotations.NotNull;
import org.jooq.Condition;
import org.jooq.Field;
import org.jooq.Param;
import org.jooq.impl.DSL;

import ch.difty.scipamato.common.TranslationUtils;
import ch.difty.scipamato.core.entity.search.IntegerSearchTerm;

/**
 * {@link SearchTermEvaluator} implementation evaluating integer searchTerms.
 *
 * @author u.joss
 */
public class IntegerSearchTermEvaluator implements SearchTermEvaluator<IntegerSearchTerm> {

    @NotNull
    @Override
    public Condition evaluate(@NotNull final IntegerSearchTerm searchTerm) {
        final String fieldName = getFieldName(searchTerm.getFieldName());
        final Field<Object> field = DSL.field(fieldName);
        final Param<Integer> value = DSL.val(searchTerm.getValue());

        switch (searchTerm.getType()) {
        case LESS_OR_EQUAL:
            return field.le(value);
        case LESS_THAN:
            return field.lt(value);
        case GREATER_OR_EQUAL:
            return field.ge(value);
        case GREATER_THAN:
            return field.gt(value);
        case RANGE:
            return field.between(value, DSL.val(searchTerm.getValue2()));
        case MISSING:
            return field.isNull();
        case PRESENT:
            return field.isNotNull();
        case INCOMPLETE:
            return DSL.falseCondition();
        default:
            return field.equal(value);
        }
    }

    private String getFieldName(final String fieldName) {
        if ("id".equals(fieldName))
            return "paper.id";
        else
            return TranslationUtils.INSTANCE.deCamelCase(fieldName);
    }
}

File: core/core-persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluator.kt
package ch.difty.scipamato.core.persistence.paper.searchorder

import ch.difty.scipamato.common.TranslationUtils.deCamelCase
import ch.difty.scipamato.core.entity.search.IntegerSearchTerm
import org.jooq.Condition
import org.jooq.impl.DSL

/**
 * [SearchTermEvaluator] implementation evaluating integer searchTerms.
 *
 * @author u.joss
 */
class IntegerSearchTermEvaluator : SearchTermEvaluator<IntegerSearchTerm?> {

    override fun evaluate(searchTerm: IntegerSearchTerm): Condition {
        val field = DSL.field(searchTerm.fieldName.sanitize())
        val value = DSL.`val`(searchTerm.value)
        return when (searchTerm.type) {
            IntegerSearchTerm.MatchType.LESS_OR_EQUAL -> field.le(value)
            IntegerSearchTerm.MatchType.LESS_THAN -> field.lt(value)
            IntegerSearchTerm.MatchType.GREATER_OR_EQUAL -> field.ge(value)
            IntegerSearchTerm.MatchType.GREATER_THAN -> field.gt(value)
            IntegerSearchTerm.MatchType.RANGE -> field.between(value, DSL.`val`(searchTerm.value2))
            IntegerSearchTerm.MatchType.MISSING -> field.isNull
            IntegerSearchTerm.MatchType.PRESENT -> field.isNotNull
            IntegerSearchTerm.MatchType.INCOMPLETE -> DSL.falseCondition()
            else -> field.equal(value)
        }
    }

    private fun String.sanitize(): String? = when {
        "id" == this -> "paper.id"
        else -> deCamelCase(this)
    }
}